web/client-api/src/main/java/io/deephaven/web/client/api/console/JsVariableDefinition.java

```diff
@@ -1,58 +1,83 @@
 import jsinterop.annotations.JsProperty;
-import jsinterop.annotations.JsType;
 import jsinterop.base.Js;
 import jsinterop.base.JsPropertyMap;
 
-import java.io.Serializable;
+public class JsVariableDefinition {
+    private final String type;
+    private final String title;
+    private final String id;
+    private final String description;
 
-@JsType(namespace = "dh")
-public class JsVariableDefinition implements Serializable {
-    private String name;
-    private String type;
-
-    @JsIgnore
     public static JsVariableDefinition from(Object definitionObject) {
         if (definitionObject instanceof JsVariableDefinition) {
             return (JsVariableDefinition)definitionObject;
         } else {
-            return new JsVariableDefinition(Js.cast(definitionObject));
+            return new JsVariableDefinition(Js.<JsPropertyMap<Object>>cast(definitionObject));
         }
     }
 
-    @JsConstructor
-    public JsVariableDefinition() {
+    public JsVariableDefinition(String type, String title, String id, String description) {
+        this.type = type;
+        this.title = title;
+        this.id = id;
+        this.description = description;
     }
 
-    @JsIgnore
-    public JsVariableDefinition(String name, String type) {
-        this();
-        this.name = name;
-        this.type = type;
+    public JsVariableDefinition(FieldInfo field) {
+        this.type = JsVariableChanges.getVariableTypeFromFieldCase(field.getFieldType().getFieldCase());
+        this.id = field.getTicket().getTicket_asB64();
+        this.title = field.getFieldName();
+        this.description = field.getFieldDescription();
     }
 
-    @JsIgnore
     public JsVariableDefinition(JsPropertyMap<Object> source) {
-        this();
-        if (source.has("name")) {
-            this.name = source.getAny("name").asString();
-        }
-
         if (source.has("type")) {
             this.type = source.getAny("type").asString();
+        } else {
+            throw new IllegalArgumentException("no type field; could not construct JsVariableDefinition");
         }
+
+        boolean hasName = source.has("name");
+        boolean hasId = source.has("id");
+        if (hasName && hasId) {
+            throw new IllegalArgumentException("has both name and id field; could not construct JsVariableDefinition");
+        } else if (hasName) {
+            this.id = source.getAny("name").asString();
+        } else if (hasId) {
+            this.id = source.getAny("id").asString();
+        } else {
+            throw new IllegalArgumentException("no name/id field; could not construct JsVariableDefinition");
+        }
+
+        this.title = "js-constructed-not-available";
+        this.description = "js-constructed-not-available";
     }
 
     @JsProperty
+    public String getType() {
+        return type;
+    }
+
+    @JsProperty @Deprecated
     public String getName() {
-        return name;
+        return title;
```

Comment:
I think this should return the `id`, as that's the unique name/equivalent

Reply:
Actually I think I'm misunderstanding something. The `name` seems to map to `title` instead of `id`.
This changes the API considerably, and IMO makes it more difficult for the client to use. The client doesn't care about these arbitrary IDs that seem to be completely random, and stores variables based on the name in that session (of which there can only be one in that scope). Within that scope (the IdeSession), it wants to get an object with that name, it should just be able to call `getTable(name)` and get that object, it shouldn't require a full definition.

Reply:
We decided to deprecate name, because it is no longer a unique identifier.

An "old variable" gets mapped like this:
id = name
title = name
description = ""
type = type

Title is just a different word for `name` and represents a human readable name for the table, but is not unique and cannot identify the variable on the server.

Reply:
> client doesn't care about these arbitrary IDs that seem to be completely random
Well, now it needs to, because it is not random it is what the server uses to identify the variable. It looks like nonsense because it is binary and is b64 encoded here.

Reply:
When the client runs code like:
```
session.runCode(`t = emptyTable(100).update("A=i")`)
```
After that completes, it's expected there's a table with the name `t` within that scope. Are you saying that is invalid? I don't understand why the client would care about the arbitrary IDs when the session/name combo should also be unique.

How are the id's generated? Will they be consistent if you restart the server?

Reply:
I guess I'm saying given the IdeSession, the variable `name` should already be unique... so why force the client to require knowledge of the id when you should be able to derive that?

And what's the point of having the `.getTable`, `.getTreeTable`, etc. methods if they still take an object just like `.getObject`? It's assumed client already knows they are tables (hence calling that method), there's no checks right now that the type actually matches the method. The session, type, and variable name are all known already, so the id should be deriveable

Reply:
The id's are persistent and consistent. For application fields they are structured like this: "a/$app_id/f/$field_name". For Query scope they are structured like "s/$field_name". These ticket patterns are not meant for the client to understand or construct on their own -- just use it to help the server identify which of these you want if they are named the same. DHE would shadow the variables exposed from the "application-script" equivalent -- which, I understand, is why remote REPL is more-or-less disabled for PQs. This solves the shadowing problem by introducing a "hybrid" mode where you can still connect to the console while in application mode.

There is a restriction that a REPL should not be able to destroy application state -- down the road when we are not sharing script sessions, this enables all connecting clients to get their own query scope namespace while still having access to the application mode exported fields.

Ideally I think the client would just keep around the JsVariableDefinition and use it to look things up (getObject, for example, already takes a JsVariableDefinition).

> And what's the point of having the .getTable, .getTreeTable, etc. methods if they still take an object just like .getObject? It's assumed client already knows they are tables (hence calling that method), there's no checks right now that the type actually matches the method. The session, type, and variable name are all known already, so the id should be deriveable

@niloc132 recommended that the client knows the type of the variable, that it should try to load it as that type -- but that is hidden from the jsapi user directly. I'm not sure I see a problem calling `getObject()` as it delegates to getTable/getTreeTable/etc. The delegate methods are there to help the jsapi catch race conditions for when a field's type changes before the UI knew about it. (i.e. load a layout that knows this tab is a Figure, but it turned out to be a Table because the application developer changed his mind but didn't change your custom layout)

Reply:
Okay, so in terms of an IdeSession - can an IdeSession have more than one $app_id then? Or is it a Query scope?
I don't understand what you mean about catching race conditions - how does it help catch a race condition? The `.getTable` function only really reads the `id` from the definition to actually retrieve the object, it doesn't check the type or title or anything. That's all just printed out as debug info.

Reply:
`getTable` calls an API method that verifies the retrieved variable is a Table and not any other kind of type. The race condition is between the API client's knowledge of the field (e.g. as a client I call `getTable` which expects to find a Table or else error) and between the server state changing (e.g. two chrome tabs are open, while one is fetching a query scope variable named `t` the other changes the variable -- one request will race the other and will cause getTable to fail in one ordering but not the other).

An IdeSession can have multiple app_id's yes. This is one thing Colin was discussing with you w.r.t. the QueryConfig pattern in DHE. An Application is a bag of fields. With this layer of abstraction, it becomes easier to think about how one might put together a dashboard that combines data from different workers.

Reply:
Okay. I think I'll be okay because I'm already calling just `getObject`... but to re-iterate - I don't see any value in exposing `.getTable`/`.getFigure`. etc and require the entire `VariableDefinition` to be passed in; or if you do, you should assert immediately that the type matches the method.